[DDP] rebuilt bucket order when find_unused_parameters=true

zhaojuanmao opened the pull request:
Differential Revision: D72437251

Enable to rebuild bucket order when find_unused_parameters=true.

It should be always better than not rebuilding bucket order when find_unused_parameters=True:

1. for cases where bucket order in the first iteration is the same as the parameter order, rebuilding bucket order will not change anything

2. for cases where bucket order in the first iteration is not the same as the parameter order, there could be two cases: 
    a. bucket order will not change after 1st iteration even the graph is dynamic and there is unused parameter, in this case, rebuilding bucket order will have performance gain 
    b. bucket order change after 1st iteration due to dynamic graph, in this case, both parameter order and bucket order in 1st iteration are not ideal, so rebuilding bucket order or not does not matter

it can help case 2.a if enabling to rebuild bucket order when find_unused_parameters=true. meanwhile it will not hurt other cases in 1 and 2.b.




cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k

rohan-varma reviewed (approved):
LGTM! Does it support the case where the set of unused parameters dynamically changes during training? In all cases, are unused parameters always in their own bucket for allreduce?

zhaojuanmao commented:
> LGTM! Does it support the case where the set of unused parameters dynamically changes during training? In all cases, are unused parameters always in their own bucket for allreduce?

yes, it will support such case, it is just the bucket order will be the same as first iteration's execution order, not the parameter order. in this case, I think it does not matter much to use parameter order or first iteration's execution order

zhaojuanmao committed:
[DDP] rebuilt bucket order when find_unused_parameters=true (#153404)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/153404

Enable to rebuild bucket order when find_unused_parameters=true.

It should be always better than not rebuilding bucket order when find_unused_parameters=True:
1.  for cases where bucket order in the first iteration is the same as the parameter order, rebuilding bucket order will not change anything
2. for cases where bucket order in the first iteration is not the same as the parameter order, there could be two cases:
   a. bucket order will not change after 1st iteration even the graph is dynamic and there is unused parameter, in this case, rebuilding bucket order will have performance gain
    b. bucket order change after 1st iteration due to dynamic graph, in this case, both parameter order and bucket order in 1st iteration are not ideal, so rebuilding bucket order or not does not matter

it can help case 2.a if enabling to rebuild bucket order when when find_unused_parameters=true. meanwhile it will not hurt other cases in 1 and 2.b.

Test Plan: unit tests

Reviewed By: fegin, rohan-varma

Differential Revision: D72437251

zhaojuanmao commented:
The PR failed for cases like GAN based models, where it could manually run multiple backward passes with DDP.prepare_for_backward() calls in each forward pass. The fix is in https://github.com/pytorch/pytorch/pull/154447. 

Considering dynamic graph may have more edge use cases than expected, will delay to push land https://github.com/pytorch/pytorch/pull/154447 as well unless we find a strong use case to benefit from it.
